File: Editor/Assets/Script/Block/BalancedKDTree.cs
using UnityEngine;
using System.Collections;


//struct NodeData{
//    NodeData(Vector3 p){
//        pos = p;
//        pData = d;
//    }
//    Vector3 pos;
//};

//class DistBuffer
//{
//    public DistBuffer(int num){
//        dists = new float[num];
//        maxIndex = 0;
//    }
//    public float getMaxDistSq(){
//        return dists[maxIndex];
//    }
//    public void update(float dist){
//        dists[maxIndex] = dist;
//        //更新最大值
//        float maxDistSq = 0;
//        maxIndex = 0;
//        for(int i = 0; i<dists.Length; i++){// 找到最远的一个
//            if(maxDistSq < dists[i]){
//                maxDistSq = dists[i];
//                maxIndex = i;
//            }
//        }
//    }
//    public int getMaxIndex(){return maxIndex;}

//    float[] dists;
//    int maxIndex;
//};

//class KDNode
//{
//    public KDNode();
//    ~KDNode();
//    public void build(NodeData** buff,int num,int depth,const AABB& aabb);
//    public void nearest(const Vector3& pos,DistBuffer& dists,std::vector<NodeData*>&results);
//    public void release();

//    int axis;
//    Real splitValue;
//     aabb;
//    bool bLeaf;
//    NodeData* data;
//    KDNode* left;
//    KDNode* right;
//};

//class KDTree
//{
//public:
//    KDTree(){
//        aabb.reset();
//    }
//    ~KDTree(){}
//    void addPoint(const Vector3& pos,void* pData){
//        buff.push_back(NodeData(pos,pData));
//        aabb.addPoint(pos);
//    }
//    void build(){
//        ptrBuffer.resize(buff.size());
//        for(unsigned int i = 0; i < buff.size(); i++){
//            ptrBuffer[i] = &(buff[i]);
//        }
//        root.build(&ptrBuffer[0],ptrBuffer.size(),0,aabb);
//    }
//    void release(){
//        root.release();
//    }

//    class Callback{
//    public:
//        Callback(){}
//        virtual ~Callback(){}
//        virtual bool isValid(void* data){return true;}
//        virtual void result(void * data);
//    };
//    void getNearestN(const Vector3& pos,Callback* cb,std::vector<NodeData*>&results,Real& maxDistSq);
//private:
//    AABB aabb;
//    KDNode root;
//    std::vector<NodeData>buff;
//    std::vector<NodeData*>ptrBuffer;
//};

File: Editor/NativePlugin/RayTrace/src/RayTrace/BalancedKDTree.cpp
#include "../stdafx.h"
#include "../core/core.h"
using namespace oo;
#include "BalancedKDTree.h"

using namespace oo;

KDNode::KDNode()
{
	left = right = NULL;
}

KDNode::~KDNode()
{

}

bool cmpx(const NodeData* a,const NodeData* b) {return a->pos.x < b->pos.x;}
bool cmpy(const NodeData* a,const NodeData* b) {return a->pos.y < b->pos.y;}
bool cmpz(const NodeData* a,const NodeData* b) {return a->pos.z < b->pos.z;}

inline void setAxisValue(Vector3& pos,int axis,Real v){
	(&(pos.x))[axis] = v;
}
inline Real getAxisValue(const Vector3& pos,int axis){
	return (&(pos.x))[axis];
}

void getAABB(const AABB& aabb,int axis,Real split,AABB& min,AABB& max)
{
	min = max = aabb;
	setAxisValue(min.vMax,axis,split);
	setAxisValue(max.vMin,axis,split);
}

void KDNode::build(NodeData** buff,int num,int depth,const AABB& aabb)
{
	if(num > 1){
		bLeaf = false;
		axis = depth % 3;
		this->aabb = aabb;
		std::sort(buff,buff + num,(axis==0)?cmpx:(axis==1)?cmpy:cmpz);
		int midIndex = num / 2;
		NodeData* midData = buff[midIndex];
		AABB leftAabb,rightAabb;
		getAABB(aabb,axis,getAxisValue(midData->pos,axis),leftAabb,rightAabb);
		left = new KDNode;
		left->build(buff,midIndex,depth + 1,leftAabb);
		if(num > 2){
			right = new KDNode;
			right->build(buff + midIndex + 1,num - midIndex - 1,depth + 1,rightAabb);
		}
		else{
			right = NULL;
		}
		data = midData;
		splitValue = getAxisValue(data->pos,axis);
	}
	else{
		bLeaf = true;
		this->aabb = aabb;
		data = buff[0];
	}
}

void KDNode::nearest(const Vector3& pos,DistBuffer& dists,std::vector<NodeData*>&results)
{
	Real distSq = (pos - data->pos).lengthSq();

	if(distSq < dists.getMaxDistSq()){
		results[dists.getMaxIndex()] = data;
		dists.update(distSq);
	}

	if(left){
		Real distToNear = getAxisValue(pos,axis) - splitValue;
		if((getAxisValue(pos,axis) < splitValue) || distToNear * distToNear < dists.getMaxDistSq()){
			left->nearest(pos,dists,results);			
		}
	}
	if(right){
		Real distToNear = getAxisValue(pos,axis) - splitValue;
		if((getAxisValue(pos,axis) > splitValue) || distToNear * distToNear < dists.getMaxDistSq()){
			right->nearest(pos,dists,results);
		}
	}
}

void KDNode::release()
{
	if(left){
		left->release();
		delete left;
		left = 0;
	}
	if(right){
		right->release();
		delete right;
		right = 0;
	}
}

void KDTree::getNearestN(const Vector3& pos,Callback* cb,std::vector<NodeData*>&results,Real& maxDistSq)
{
	DistBuffer distBuff(results.size());
	root.nearest(pos,distBuff,results);
	maxDistSq = distBuff.getMaxDistSq();
}